modules/core/src/main/java/org/apache/ignite/internal/util/future/IgniteFutureImpl.java

```diff
@@ -68,7 +80,10 @@ public IgniteFutureImpl(IgniteInternalFuture<V> fut) {
     @Override public void listen(IgniteInClosure<? super IgniteFuture<V>> lsnr) {
         A.notNull(lsnr, "lsnr");
 
-        fut.listen(new InternalFutureListener(lsnr));
+        if (defaultExecutor != null && !isDone())
+            fut.listen(new InternalFutureListener(new AsyncFutureListener<>(lsnr, defaultExecutor)));
```

Comment:
So the `defaultExecutor` overwrites explicitly set `lsnr`? Does not seem right to me.

Reply:
We don't overwrite `lsnr`. We wrap `lsnr` into `AsyncFutureListener` which uses the provided executor to run the delegate.